lib/Sema/TypeCheckStmt.cpp

```diff
@@ -1269,8 +1270,13 @@ Stmt *StmtChecker::visitBraceStmt(BraceStmt *BS) {
       break;
 
     TC.typeCheckDecl(SubDecl, /*isFirstPass*/false);
+    if (SubDecl->isInvalid())
```

Comment:
I agree that it is more principled to never leave null types around, but I think it still makes sense to communicate the error out and not do e.g. the misc. diagnostics that normally do not do when type checking fails, and the only way that I currently know of to do that is check `isInvalid()` on the decl, or try to dig around for an ErrorType in the expression. For example for the test case from `SR-5666` (which I'm going to add to the PR) we end up with this. We end up crashing on the string_literal_exprs here:

```
(pattern_binding_decl
  (pattern_named type='<<error type>>' 'b')
  (interpolated_string_literal_expr type='<<error type>>' location=/tmp/ppu.swift:8:15 range=[/tmp/ppu.swift:8:15 - line:8:15]
    (string_literal_expr type='<null>' encoding=utf8 value="" builtin_initializer=**NULL** initializer=**NULL**)
    (paren_expr type='<null>'
      (declref_expr type='<<error type>>' location=/tmp/ppu.swift:8:18 range=[/tmp/ppu.swift:8:18 - line:8:18] decl=ppu.(file).testSR5666(cs:).explicit closure discriminator=0.a@/tmp/ppu.swift:7:11 function_ref=unapplied))
    (string_literal_expr type='<null>' encoding=utf8 value="" builtin_initializer=**NULL** initializer=**NULL**)))
```